### DISCUSSION TITLE
Proposal: Rename default branch to main

### DISCUSSION THREAD

**Discussion by colby-swandale** [ruby/rubygems#5145] [Ideas]
## Abstract

I wish to propose renaming the default branch from `master` to `main`. The larger developer community has been moving away from non-inclusive terms such as master, slave, blacklist, whitelist, and I wish for RubyGems & RubyGems.org to do the same.

Other examples of open source projects renaming their default branch main:

 * https://github.com/rails/rails/issues/40214
 * https://github.com/rails/rails/pull/40254
 * https://make.wordpress.org/core/2020/06/18/proposal-update-all-git-repositories-to-use-main-instead-of-master/

This change is relatively minimal thanks to Github natively supporting renaming the default branch to `main` and will do most of the work for us automatically, such as adding automatic redirects & updating existing Pull Requests to point to `main`. But this will require some manual steps.

You can find more information at: https://github.com/github/renaming.

## Steps Required

As mentioned, Github automates most of the needed steps to switch the default branch to `main`, but some steps still need to be performed manually by maintainers before & after the rename.

The following list covers steps that to my knowledge are required to do, but may not be indicative of covering everything.

#### Github Actions

Before renaming the default branch in Github, the Github Action Workflows will need to be amended to mirror the `main` and `master` branches. So that we can test pushing a commit to `main`, while developers continue to use `master`.

#### Update branch protection rules

We currently use branch protection to prevent merging Pull Requests that do not pass the status checks. This rule will need to be updated to use `main` after the default branch changes.

#### Local Migration

After the rename, developers will need to migrate their local development environment to `main`. This migration can be done by renaming the `master` branch to `main` and updating the remote to `origin/main`

```
$ git checkout master
$ git branch -m main
$ git branch -u origin/main
```

#### Remove/Rename references to master branch

Update Github actions to remove any branch filtering on the `master` branch and rename any references to `master` in the documentation & source code.

## Renaming the default branch

The default branch can be renamed in Github, which is documented in [doc.github.com](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch)

## Questions

* What impacts would this change cause downstream, such as Ruby-src, Truffleruby & JRuby?
  * How best do we communicate this change to maintainers?

* Are any scripts/tooling affected by this proposed change?
* Is the workflow in bundler-site affected by this proposed change?

**Comment by simi** [comment]
IMHO we should just copy Ruby repo standards and discuss overall renaming of all Ruby repos (including std. libs - RubyGems) at https://bugs.ruby-lang.org/.

**Comment by hsbt** [comment]
Ruby core followed [Matz's opinion](https://bugs.ruby-lang.org/issues/17828#change-91702). I also agreed him.

**Comment by deivid-rodriguez** [comment]
I'm in favor of doing this.

**Comment by deivid-rodriguez** [comment]
> What impacts would this change cause downstream, such as Ruby-src, Truffleruby & JRuby?

It does affect syncing with ruby-core because current scripts hardcode "master" as the default branch. They would need to be updated to infer the default branch for each repository instead of hardcoding it.

> Are any scripts/tooling affected by this proposed change?

The scripts that generate the changelog will be affected, and also GitHub workflows.

> Is the workflow in bundler-site affected by this proposed change?

Most likely yes.

**Comment by deivid-rodriguez** [comment]
> We currently use branch protection to prevent merging Pull Requests that do not pass the status checks. This rule will need to be updated to use main after the default branch changes.

I skimmed through https://github.com/github/renaming and apparently GitHub does this now automatically too 🎉.

**Comment by deivid-rodriguez** [comment]
I moved this to a discussion, it fits better here in my opinion.

**Comment by indirect** [comment]
I'm also in favor of changing the repo's default branch to `main`. Following best practices is something that has been important to the entire Bundler team for many years. In the beginning, those best practices included having an extensive test suite, limiting breaking changes, and having an extensive documentation website. Over the last 5 or so years, those best practices have grown to include using an automatic linter, having a code of conduct, and using the default git branch `main`.

Personally, I want Bundler to continue to be an example of current best practices, as much as possible, so I am strongly in favor of making this change.

**Reply by simi** [reply]
I think the major problems are actually technical ones. With limited time we have to work on the projects this is not most critical problem and thus not on top of the priority list. Anyone is welcomed to take a look to summary the impact (mainly synchronize with Ruby side) and help with transition.

**Reply by deivid-rodriguez** [reply]
Indeed the ruby-core synchronization script hardcodes "master" assuming all default repositories have their default branch named master. So the `tool/sync_default_gems.rb` script in ruby/ruby needs to be made generic first.

**Comment by deivid-rodriguez** [comment]
Quick update, I renamed [bundler-site](https://github.com/rubygems/bundler-site/) default branch to main. Also I noticed that:

* The default gem import script now supports different default branch names, so we'd just need to update the script to specify our new default branch, like [prism does](https://github.com/ruby/ruby/blob/4a06ef98bfd480a3d724b16c2d7da071e373a69c/tool/sync_default_gems.rb#L45).
* We have a bunch of source code links explicitly including "master" in both rubygems/rubygems and rubygems/bundler-site, however, those links gracefully redirect to the same links using "main" when the default branch is renamed, so that should be handled gracefully.

Overall, this seems less trouble than I initially thought.
